esi_leap/db/sqlalchemy/api.py

```diff
@@ -370,17 +472,94 @@ def resource_verify_availability(r_type, r_uuid, start, end):
+         (end >= models.Lease.end_time))
     )).first()
 
     if lease_conflict:
         raise exception.ResourceTimeConflict(
             resource_uuid=r_uuid,
             resource_type=r_type)
+
+    # check conflict with ownership changes
+    if not is_owner_change:
+        # check_resource_admin will have been called earlier
+        # for leases and offers
+        return
+
+    oc_query = model_query(models.OwnerChange)
+
+    ocs = oc_query.with_entities(
+        models.OwnerChange.start_time, models.OwnerChange.end_time).\
+        filter((models.OwnerChange.resource_uuid == r_uuid),
+               (models.OwnerChange.resource_type == r_type),
+               (models.OwnerChange.status.in_([statuses.CREATED,
+                                               statuses.ACTIVE])))
+
+    oc_conflict = ocs.filter((
+        ((start >= models.OwnerChange.start_time) &
+         (start < models.OwnerChange.end_time)) |
+
+        ((end > models.OwnerChange.start_time) &
+         (end <= models.OwnerChange.end_time)) |
+
+        ((start <= models.OwnerChange.start_time) &
+         (end >= models.OwnerChange.end_time))
+    )).first()
+
+    if oc_conflict:
+        raise exception.ResourceTimeConflict(
+            resource_uuid=r_uuid,
+            resource_type=r_type)
+
+
+def resource_check_admin(resource_type, resource_uuid,
+                         start_time, end_time,
+                         default_admin_project_id, project_id):
+    # check if time period straddles an owner changes
+    ocs_conflicts = model_query(models.OwnerChange).with_entities(
+        models.OwnerChange.start_time, models.OwnerChange.end_time).\
+        filter((models.OwnerChange.resource_uuid == resource_uuid),
+               (models.OwnerChange.resource_type == resource_type),
+               (models.OwnerChange.status.in_([statuses.CREATED,
+                                               statuses.ACTIVE])))
+
+    ocs_conflicts = ocs_conflicts.filter((
+        ((start_time >= models.OwnerChange.start_time) &
+         (start_time < models.OwnerChange.end_time) &
+         (end_time > models.OwnerChange.end_time)) |
+
+        ((start_time < models.OwnerChange.start_time) &
+         (end_time > models.OwnerChange.start_time) &
+         (end_time < models.OwnerChange.end_time)) |
+
+        ((start_time <= models.OwnerChange.start_time) &
+         (end_time >= models.OwnerChange.end_time))
+    ))
+
+    if ocs_conflicts.count() > 0:
+        return False
+
+    # check if time period encompasses a single owner change
+    filters = {
+        'resource_type': resource_type,
+        'resource_uuid': resource_uuid,
+        'start_time': start_time,
+        'end_time': end_time,
+        'status': [statuses.CREATED, statuses.ACTIVE]
+    }
+    ocs = owner_change_get_all(filters)
+
+    if ocs.count() > 1:
+        # shouldn't happen, but...
+        return False
+    if ocs.count() == 1:
+        return project_id == ocs[0].to_owner_id
```

Comment:
I don't understand here... Does resource_check_admin() check if the given project is the admin project?

Reply:
I see! Thanks!